File: src/CPF.js
/**
 * Create, validate and format CPFs
 *
 * @author Roger Lima (rogerlima@outlook.com)
 * @date 21/feb/2015
 * @update 15/apr/2016
 */
/* jshint laxbreak: true, bitwise: false */

( function( window ) {
'use strict';

/**
 * @object CPF
*/
var CPF = {};

/**
 * Gets the verifier of a given set of numbers
 *
 * @param {string} Group of nine random numbers
 * @return {number}
 */
function get_verifier( seed ) {
	for ( var i = 0, verifier = 0; i < seed.length; verifier += seed[ i ] * ( seed.length - i++ + 1 ) );

	return ( verifier % 11 ) < 2 ? 0 : 11 - ( verifier % 11 );
}

/**
 * Format a CPF
 *
 * @param {string} cpf CPF to format
 */
CPF.format = function( cpf ) {
	if ( typeof cpf !== 'string' )
		throw 'CPF.validate: "cpf" argument need be a string';

	return cpf.replace( /[^\d]/, '' ).substr( 0, 11 )
		.replace( /(\d{3})(?=\d{3})/g, '$1.' )
		.replace( /(\d{2})$/, "-$1" );
};


/**
 * Create a valid CPF
 *
 * @param {boolean} Format or not the CPF
 * @param [{string|number}] Tax Region. If empty, random region is setted;
 * Tax region possibly values
	1 (Centro-Oeste, Tocantins)
	2 (Norte -Tocantins)
	3 (Maranhão, Piauí, Ceará)
	4 (Rio Grande do Norte, Pernambuco, Paraíba)
	5 (Bahia, Sergipe)
	6 (Minas Gerais)
	7 (Rio de Janeiro, Espírito Santo)
	8 (São Paulo)
	9 (Paraná, Santa Catarina)
	0 or 10 (Rio Grande do Sul)
 *
 * @return {string}
 */
CPF.generate = function( format, region ) {
	var first_verifier,
		seed = ( ( ( Math.random() * 9e8 ) >> 0 ) + '' );

	while ( seed.length != 9 )
		seed = "0" + seed;

	if ( typeof region !== 'undefined' && region !== '' ) {
		if ( region >= 0 && region <= 10 )
			seed = seed.substr( 0, 8 ) + ( region == 10 ? 0 : region );
		else
			throw 'CPF.generate: the "region" range need be between 0 and 10';
	}

	first_verifier = get_verifier( seed );

	return ( format
		? seed.replace( /(\d{3})(?=\d{3})/g, '$1.' ) + '-'
		: seed
	) + first_verifier + get_verifier( seed + first_verifier );
};

/**
 * Validate a CPF
 *
 * @param {string} CPF to validate
 * @return {boolean}
 */
CPF.validate = function( cpf ) {
	var i, same_numbers, first_verifier;

	if ( typeof cpf === 'undefined' )
		throw 'CPF.validate: missing "cpf" argument';
	else if ( typeof cpf !== 'string' )
		throw 'CPF.validate: "cpf" argument need be a string';

	cpf = cpf.replace( /[^\d]/g, '' );
	first_verifier = get_verifier( cpf.substr( 0, 9 ) );

	// Prevents false positive (111.111.111-11, ...)
	for ( i = 0, same_numbers = 0; i < cpf.length; ( cpf[ i++ ] === cpf[ i ] ) && same_numbers++ );

	return same_numbers < 10
		&& cpf.length === 11
		&& cpf[ cpf.length - 2 ] == first_verifier
		&& cpf[ cpf.length - 1 ] == get_verifier( cpf.substr( 0, 9 ) + first_verifier );
};

window.CPF = CPF;

}( window ) );

File: src/CPF.cpp
#include "CPF.hpp"

int CPF::get_verifier( std::string seed ) {
	size_t i = 0,
		verifier = 0;

	while ( i < seed.size() )
		verifier += atoi( std::string( 1, seed[ i ] ).c_str() ) * ( seed.size() - i++ + 1 );

	return ( verifier % 11 ) < 2 ? 0 : 11 - ( verifier % 11 );
};

std::string CPF::format( std::string cpf ) {
	return std::regex_replace(
		std::regex_replace( 	
			std::regex_replace( 
				cpf, 
				std::regex::basic_regex( "[^\\d]" ), "" ).substr( 0, 11 ), 
			std::regex::basic_regex( "(\\d{3})(?=\\d{3})" ), "$1." 
		),
		std::regex::basic_regex( "(\\d{2})$" ), "-$1" 
	);
};

std::string CPF::generate( bool format, int region ) {
	std::random_device rd;
	std::uniform_real_distribution< double > randomize( 0, 1 );

	std::string seed = std::to_string( long ( ( randomize( rd ) * 9e8 ) ) );

	while ( seed.size() != 9 )
		seed = "0" + seed;

	if ( region >= 0 && region <= 10 )
		seed = seed.substr( 0, 8 ) + ( region == 10 ? "0" : std::to_string( region ) );

	std::string first_verifier = std::to_string( get_verifier( seed ) );

	return ( format
		? std::regex_replace( seed, std::regex::basic_regex( "(\\d{3})(?=\\d{3})" ), "$1." ) + "-"
		: seed
	) + first_verifier + std::to_string( get_verifier( seed + first_verifier ) );
};

bool CPF::validate( std::string cpf ) {
	size_t i,
		same_numbers;

	cpf = std::regex_replace( cpf, std::regex::basic_regex( "[^\\d]" ), "" );

	std::string base_CPF = cpf.substr( 0, 9 );
	char first_verifier = std::to_string( get_verifier( base_CPF ) )[ 0 ],
		second_verifier = std::to_string( get_verifier( base_CPF + first_verifier ) )[ 0 ];

	// Prevents false positive (111.111.111-11, ...)
	for ( i = same_numbers = 0; i < cpf.size(); ( ( cpf[ i ] == cpf[ i + 1 ] ) && same_numbers++ ), i++ );

	return same_numbers < 10
		&& cpf.size() == 11 
		&& first_verifier == cpf[ cpf.size() - 2 ]
		&& second_verifier == cpf[ cpf.size() - 1 ];
};